### PR TITLE
Guardrail in Imp genesis

### LINKED ISSUE #4726: Add guardrails script to the Conway genesis  in the imp spec
All of our guardrails script rely on adding a script. This is not the case for mainnet, guardrails script is already present.

We need to implement a basic guardrails script and use it in the Conway genesis. Also we need to adjust all the tests that have been disabled for bootstrap phase in #4713

### PR DESCRIPTION
**Pull request by teodanciu**
# Description

Initialises ImpTests with the current constitution, as published in the [Cardano Operations Book](https://book.world.dev.cardano.org/env-mainnet.html) 

Closes https://github.com/IntersectMBO/cardano-ledger/issues/4726

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages.
      **_New section is never added with the code changes._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated.
      **_If you change the bounds in a cabal file, that package itself must have a version increase._** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [ ] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 1]
```diff
@@ -300,29 +302,30 @@ instance
           { anchorUrl = errorFail $ textToUrl 128 "https://cardano-constitution.crypto"
           , anchorDataHash = hashAnchorData (AnchorData "Cardano Constitution Content")
           }
+      constitutionHash = hashPlutusScript guardrail
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs]: It is not really hash of a constitution. Anchor is that
```suggestion
      guardrailScriptHash = hashPlutusScript guardrail
```

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/Guardrail.hs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Short as SBS (toShort)
+import qualified Data.Text as Text (Text)
+import GHC.Stack
+
+decodeHexPlutus :: HasCallStack => Text.Text -> Plutus l
+decodeHexPlutus = either (error . show) (Plutus . PlutusBinary . SBS.toShort) . decodeFullFromHexText @ByteString
+
+guardrail :: Plutus 'PlutusV3
+guardrail =
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/Guardrail.hs]: That is just how it is usually called.
```suggestion
guardrailScript :: Plutus 'PlutusV3
guardrailScript =
```

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/Guardrail.hs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  Plutus (..),
+  PlutusBinary (..),
+ )
+import Data.ByteString (ByteString)
+import qualified Data.ByteString.Short as SBS (toShort)
+import qualified Data.Text as Text (Text)
+import GHC.Stack
+
+decodeHexPlutus :: HasCallStack => Text.Text -> Plutus l
+decodeHexPlutus = either (error . show) (Plutus . PlutusBinary . SBS.toShort) . decodeFullFromHexText @ByteString
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/Guardrail.hs]: I think it would be better to move this function to `Test.Cardano.Ledger.Plutus` due to its generality.

Also, I don't think it is necessary to decode into a `ByteString` first. It should work equivalently well by decoding directly into a `ShortByteString`
```suggestion
decodeHexPlutus :: HasCallStack => Text.Text -> Plutus l
decodeHexPlutus = either (error . show) (Plutus . PlutusBinary) . decodeFullFromHexText @ByteString
```

By the way, the reason why decoding directly into `Plutus l` did not work for you, because that is not the decoder we use for dealing with scripts on mainnet:
https://github.com/IntersectMBO/cardano-ledger/blob/373c927b6fca77bd997624fd84567e280175a561/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Scripts.hs#L582-L584

Which has to do with the fact that tag is reused for native scripts and plutus scripts at the same time.

In other words decoder for `Plutus l` encodes the language as a tag (which is needed for encoding `PlutusWithContext` for debugging), while decoder for scripts on chain (encoded transactions) uses a different tag for this purpose. The script itself never actually encodes the plutus version, since binary representation of a plutus script could often be used for different plutus versions.

File: eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs [hunk 4]
```diff
@@ -897,9 +906,23 @@ submitParameterChange ::
...
+  PParamsUpdate era ->
+  ImpTestM era (GovAction era)
+mkParameterChangeGovAction parent ppu =
+  ParameterChange (GovPurposeId <$> parent) ppu <$> getGovPolicy
+
+mkMinFeeUpdateGovAction ::
+  ConwayEraImp era =>
+  StrictMaybe (GovActionId (EraCrypto era)) -> ImpTestM era (GovAction era)
+mkMinFeeUpdateGovAction p = do
+  minFeeValue <- uniformRM (30, 1000)
```
Comment (lehins) [inline eras/conway/impl/testlib/Test/Cardano/Ledger/Conway/ImpTest.hs:907]: Nice! Randomness is always welcomed in testing :smiley:

**Review by lehins — APPROVED** [review]
This looks great. I could only suggest a few minor improvements.

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/Guardrail.hs (hunk 3 above)
Comment (teodanciu) [reply libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus/Guardrail.hs]: Thanks for explanation! 
Moved the function to Plutus module and deserialized directly as ShortByteString as you suggested :ok_hand:

File: libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus.hs [hunk 5]
```diff
@@ -127,3 +132,10 @@ alwaysSucceedsPlutus n = Plutus (PlutusBinary (P.alwaysSucceedingNAryFunction n)
 
 alwaysFailsPlutus :: Natural -> Plutus l
 alwaysFailsPlutus n = Plutus (PlutusBinary (P.alwaysFailingNAryFunction n))
+
+decodeHexPlutus :: HasCallStack => Text.Text -> Plutus l
+decodeHexPlutus =
+  either
+    (error . show)
+    (Plutus . PlutusBinary)
+    . decodeFullFromHexText @ShortByteString
```
Comment (lehins) [inline libs/cardano-ledger-core/testlib/Test/Cardano/Ledger/Plutus.hs]: Type annotation shouldn't be necessary. Type inference should be able to figure it out.

```suggestion
    . decodeFullFromHexText
```
